lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1424 @@
+          StringAttr::get(&getContext(), Twine(neededName.getValue()) +
+                                             Twine("_") + Twine(suffix++));
+    } while (llvm::is_contained(portNames, newName));
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Creating uniquified port " << newName << "\n");
+    domain.newPortName = newName;
+    return;
+  }
+
+  // At this point we know that there is no such port, and we can safely
+  // create one as needed.
+  LLVM_DEBUG(llvm::dbgs() << "- Creating new port " << neededName << "\n");
+  domain.newPortName = neededName;
+}
+
+//===----------------------------------------------------------------------===//
+// Async Reset Implementation
+//===----------------------------------------------------------------------===//
+
+/// Implement the async resets gathered in the pass' `domains` map.
+LogicalResult InferResetsPass::implementAsyncReset() {
+  for (auto &it : domains)
+    if (failed(implementAsyncReset(cast<FModuleOp>(it.first),
+                                   it.second.back().first)))
+      return failure();
+  return success();
+}
+
+/// Implement the async resets for a specific module.
+///
+/// This will add ports to the module as appropriate, update the register ops in
+/// the module, and update any instantiated submodules with their corresponding
+/// reset implementation details.
+LogicalResult InferResetsPass::implementAsyncReset(FModuleOp module,
+                                                   ResetDomain &domain) {
+  LLVM_DEBUG(llvm::dbgs() << "Implementing async reset for " << module.getName()
+                          << "\n");
+
+  // Nothing to do if the module was marked explicitly with no reset domain.
+  if (!domain.reset) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Skipping because module explicitly has no domain\n");
+    return success();
+  }
+
+  // If needed, add a reset port to the module.
+  Value actualReset = domain.existingValue;
+  if (domain.newPortName) {
+    ModulePortInfo portInfo{.name = domain.newPortName,
+                            .type = AsyncResetType::get(&getContext()),
+                            .direction = Direction::Input,
+                            .loc = domain.reset.getLoc()};
+    module.insertPorts({{0, portInfo}});
+    actualReset = module.getArgument(0);
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Inserted port " << domain.newPortName << "\n");
+  }
+  assert(actualReset);
+  LLVM_DEBUG({
+    llvm::dbgs() << "- Using ";
+    if (auto blockArg = actualReset.dyn_cast<BlockArgument>())
+      llvm::dbgs() << "port #" << blockArg.getArgNumber() << " ";
+    else
+      llvm::dbgs() << "wire/node ";
+    llvm::dbgs() << getResetName(actualReset) << "\n";
+  });
+
+  // Update the operations in the module.
+  SmallVector<Operation *> deleteOps;
+  SmallVector<std::pair<Value, Value>> connects;
+  module.walk([&](Operation *op) {
+    implementAsyncReset(op, module, actualReset, deleteOps, connects);
+  });
```

Comment:
`implementAsyncReset` inserts and removes operations (possibly both before and after the current `op` in the IR). This seems very dangerous to me, and possibly causes nodes generated by an earlier `implementAsyncReset` to be visited by the walk further down (which isn't intended).

Any thoughts on better ways to do this? I saw `IRRewriter` in some other code, but wasn't sure if that defers IR mutation until destruction. Is that a thing?

Reply:
I don't think that's a thing.  There are various ways to defer insertion, e.g. create nodes but don't insert them anywhere (put them in a vector instead) and insert all at once later.  Generally it is better to split passes into atomic transactions of the form "analysis some stuff; change/insert some stuff if successful" when those transformations compose with each other.  Why wouldn't the transforms compose here?

Reply:
Here's another way to look at it: do a walk of the module to collect all the mems and regs with something like this.  Now you have a list of resetables to transform, and are done walking the IR.

Given that list, in a separate loop, foreach the list and "implementAsyncReset" on them.  This should be a "clean" transformation that touches/deletes random nodes and the specified reg/mem, but does not transform any other reg/mems.  Given this, the list you built should never be invalidated.  This also hopefully allows you to pull `deleteOps` and `connects ` into the `implementAsyncResets` implementation.

Reply:
That is a useful pattern, makes a lot of sense. I've changed things around to first collect all resettables into a list that can then be updated cleanly as you suggest.